homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,294 @@
+            "scope": " ".join(scopes),
+            "nonce": nonce,
+        }
+
+        return str(
+            URL(self._discovery_document["authorization_endpoint"]).with_query(query)
+        )
+
+    @property
+    def support_mfa(self) -> bool:
+        """Return whether multi-factor auth supported by the auth provider."""
+        return False
+
+    async def async_get_or_create_credentials(
+        self, flow_result: Dict[str, str]
+    ) -> Credentials:
+        """Get credentials based on the flow result."""
+        subject = flow_result["sub"]
+
+        for credential in await self.async_credentials():
+            if credential.data["sub"] == subject:
+                return credential
+
+        return self.async_create_credentials(flow_result)
+
+    async def async_user_meta_for_credentials(
+        self, credentials: Credentials
+    ) -> UserMeta:
+        """Return extra user metadata for credentials.
+
+        Will be used to populate info when creating a new user.
+        """
+        email = credentials.data["email"]
+        if "name" in credentials.data:
+            name = credentials.data["name"]
+        else:
+            match = re.match(r"[^@]+", email)
+            if match:
+                name = str(match.group(0))
+            else:
+                name = str(email)
+
+        return UserMeta(name=name, is_active=True)
+
+
+class OpenIdLoginFlow(LoginFlow):
+    """Handler for the login flow."""
+
+    external_data: str
+    nonce: str
+
+    async def async_step_init(
+        self, user_input: Optional[Dict[str, str]] = None
+    ) -> Dict[str, Any]:
+        """Handle the step of the form."""
+        return await self.async_step_authenticate()
+
+    async def async_step_authenticate(
+        self, user_input: Optional[Dict[str, str]] = None
+    ) -> Dict[str, Any]:
+        """Authenticate user using external step."""
+
+        provider = cast(OpenIdAuthProvider, self._auth_provider)
+
+        if user_input:
+            self.external_data = str(user_input)
+            return self.async_external_step_done(next_step_id="authorize")
+        self.nonce = secrets.token_hex()
+        url = await provider.async_generate_authorize_url(self.flow_id, self.nonce)
+        return self.async_external_step(step_id="authenticate", url=url)
+
+    async def async_step_authorize(
+        self, user_input: Optional[Dict[str, str]] = None
+    ) -> Dict[str, Any]:
+        """Authorize user received from external step."""
+
+        provider = cast(OpenIdAuthProvider, self._auth_provider)
+        try:
+            token = await provider.async_retrieve_token(self.external_data)
+            result = await provider.async_validate_token(token, self.nonce)
+        except InvalidAuthError:
+            return self.async_abort(reason="invalid_auth")
+        return await self.async_finish(result)
+
+
+class OpenIdCallbackView(HomeAssistantView):
```

Comment:
This is almost identical to OAuth2AuthorizeCallbackView, the problem being that that view is fixed to config flow. I could possibly let the state contain flow_type.